### DISCUSSION TITLE
post_233852

### DISCUSSION THREAD

**Discussion by cincuranet** [cincuranet/blog-comments#1157] [Comments]
imported 

https://www.tabsoverspaces.com/id/233852

**Comment by cincuranet** [comment]
Imported comment written by _Meziantou_ on _2021-03-09T03:12:37_:

<p>That's why you should use <a href="https://www.meziantou.net/writing-a-language-agnostic-roslyn-analyzer-using-ioperation.htm" rel="nofollow noopener" title="https://www.meziantou.net/writing-a-language-agnostic-roslyn-analyzer-using-ioperation.htm">the IOperation model</a> to abstract the syntax when possible. In this case it is an <code>IUsingOperation</code> with the property <code>IsAsynchronous</code> equals to true. You don't need to know how the user write it, you just care about the semantic. Also, IOperation gives direct access to the semantic model information and the syntax node if needed.<br>You can check this example <a href="https://github.com/meziantou/Meziantou.Analyzer/blob/fe2b3e607bb5005bf6a675922f8682df84440583/src/Meziantou.Analyzer/Rules/UseConfigureAwaitAnalyzer.cs#L167" rel="nofollow noopener" title="https://github.com/meziantou/Meziantou.Analyzer/blob/fe2b3e607bb5005bf6a675922f8682df84440583/src/Meziantou.Analyzer/Rules/UseConfigureAwaitAnalyzer.cs#L167">https://github.com/mezianto...</a></p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2021-03-09T20:23:33_:

<p>My ConfigureAwaitChecker is also a standalone library and hence I'm operating on the whole syntax tree (in the library, the analyzer just uses it). Although I'm not sure anybody uses that, most people probably use the analyzer and the `IOperation` could be easier way.</p><p>Anyway, this is a nice brain training for me.</p>

**Comment by martyt** [comment]
Any update on this issue? Just installed your NuGet package (v5.0.0.1) and it's complaining about every `await using` in my application. I'd really like to use the package, as our devs have a habit of forgetting `.ConfigureAwait(false)`, but it's not going to be usable like this...

**Comment by cincuranet** [comment]
@martyt What issue? Can you share some code where it incorrectly produces warning?

**Comment by martyt** [comment]
@cincuranet Here's a minimal .NET 6 console app repro - using `ConfigureAwaitChecker.Analyzer` v5.0.0.1, in VS2022 Professional v17.2.5
```

Console.WriteLine("Hello, World!");

var y = await DoSomethingElseAsync().ConfigureAwait(false);

async Task<int> DoSomethingElseAsync()
{
    var foo = new Foo();
    await using var s = await foo.GetStreamAsync().ConfigureAwait(false);

    return 0;
}

public class Foo
{
    public async Task<Stream> GetStreamAsync()
    {
        return await this.AllocateStreamAsync().ConfigureAwait(false);
    }

    private async Task<Stream> AllocateStreamAsync()
    {
        return await Task.FromResult(new MemoryStream()).ConfigureAwait(false);
    }
}
```
I'm getting an "intermittent" CAC001 error on the `await using` call at line 8. And by "intermittent" I mean sometimes it's there, sometimes it's not - see the following screen captures: 
![image](https://user-images.githubusercontent.com/3867474/176260658-043c9c94-c11f-483e-8374-50aed7519511.png)
After doing nothing but deleting the blank line after `Console.WriteLine()` I get this
![image](https://user-images.githubusercontent.com/3867474/176260815-d044d17a-456f-4663-8a1f-003fffb6a676.png)
And then sometimes it disappears completely and comes back seemingly at random based on unrelated code changes.

**Comment by martyt** [comment]
@cincuranet Would you prefer that I open an issue on your Github repo? I didn't realize this was for blog posts only!

**Comment by cincuranet** [comment]
Yes.
